modules/compiler/utils/source/manual_type_legalization_pass.cpp

```diff
@@ -31,25 +32,39 @@ using namespace llvm;
 
 PreservedAnalyses compiler::utils::ManualTypeLegalizationPass::run(
     Function &F, FunctionAnalysisManager &FAM) {
-  auto &TTI = FAM.getResult<TargetIRAnalysis>(F);
-
   auto *HalfT = Type::getHalfTy(F.getContext());
   auto *FloatT = Type::getFloatTy(F.getContext());
-
-  // Targets where half is a legal type do not need this pass. Targets where
-  // half is promoted using "soft promotion" rules also do not need this pass.
-  // We cannot reliably determine which targets these are, but that is okay, on
-  // targets where this pass is not needed it does no harm, it merely wastes
-  // time.
+  auto *DoubleT = Type::getDoubleTy(F.getContext());
+
+  // Targets where half is a legal type, and targets where half is promoted
+  // using "soft promotion" rules, are assumed to implement basic operators
+  // correctly. We cannot reliably determine which targets use "soft promotion"
+  // rules so we hardcode the list here.
+  //
+  // FMA is promoted incorrectly on all targets without hardware support, even
+  // when using "soft promotion" rules; only targets that have native support
+  // implement it correctly at the moment.
+  //
+  // Both for operators and FMA, whether the target implements the operation
+  // correctly may depend on the target feature string. We ignore that here for
+  // simplicity.
   const llvm::Triple TT(F.getParent()->getTargetTriple());
-  if (TTI.isTypeLegal(HalfT) || TT.isX86() || TT.isRISCV()) {
+
+  auto &TTI = FAM.getResult<TargetIRAnalysis>(F);
+  const bool HaveCorrectHalfOps =
```

Comment:
won't some ARM target support this? Also is RISCV a guarantee  if it doesn't have hardware feature?